File: web/src/sim/military/Blockade.ts
import type { ProvinceGraph } from '~/sim/data/WorldFile'
import { addF32, f32, mulF32, subF32 } from '~/sim/determinism/float32'
import type { Relations } from '~/sim/diplomacy/Relation'
import type { Army } from '~/sim/military/Army'
import { Domain } from '~/sim/military/UnitClass'
import { NO_OWNER } from '~/sim/world/ProvinceStore'
import type { WorldState } from '~/sim/world/WorldState'

export interface BlockadeReport {
  readonly province: number
  readonly blockader: number
  readonly victim: number
  readonly fleetStrength: number
}

/* Armada yang memblokade harus melebihi apa pun yang menjaga lintasannya, atau
   satu kapal patroli bisa mencekik sebuah garis pantai. */
export const STRENGTH_RATIO = f32(2)

export const PRODUCTION_PENALTY = f32(0.5)
export const MORALE_PENALTY_PER_DAY = f32(0.03)
export const SIEGE_PROGRESS_PER_DAY = 3

const MORALE_FLOOR = f32(0.05)

/*
 * Armada yang duduk di tautan laut sebuah provinsi pesisir memutusnya dari
 * dunia. Di kepulauan inilah satu-satunya pengepungan yang ada: graf lautnya
 * cukup padat sehingga tidak ada satu pun provinsi darat yang jadi titik
 * cekik, jadi pengepungan Indonesia harus dilakukan di laut atau tidak sama
 * sekali.
 */
export class Blockade {
  readonly #blockaded: Uint8Array
  readonly #blockader: Uint16Array
  readonly #reports: BlockadeReport[] = []

  constructor(
    private readonly world: WorldState,
    private readonly relations: Relations,
    private readonly sea: ProvinceGraph,
  ) {
    this.#blockaded = new Uint8Array(world.provinces.count)
    this.#blockader = new Uint16Array(world.provinces.count)
  }

  get reports(): readonly BlockadeReport[] {
    return this.#reports
  }

  isBlockaded(province: number): boolean {
    return this.#blockaded[province] !== 0
  }

  blockaderOf(province: number): number {
    return this.#blockader[province]!
  }

  productionMultiplierFor(province: number): number {
    return this.#blockaded[province] !== 0 ? subF32(1, PRODUCTION_PENALTY) : 1
  }

  /* Pesisir berarti ia punya lintasan laut sama sekali. Provinsi pedalaman
     tidak bisa diblokade sekuat apa pun armadanya, dan itulah yang membuat
     memegang pedalaman bernilai sesuatu di sebuah kepulauan. */
  isCoastal(province: number): boolean {
    return this.sea.neighboursOf(province).length > 0
  }

  recompute(armies: ReadonlyMap<number, Army>): void {
    this.#blockaded.fill(0)
    this.#blockader.fill(NO_OWNER)
    this.#reports.length = 0

    const fleetsByProvince = new Map<number, Map<number, number>>()

    for (const army of armies.values()) {
      if (army.isDestroyed || !carriesFleet(army)) continue

      let byNation = fleetsByProvince.get(army.province)
      if (byNation === undefined) {
        byNation = new Map<number, number>()
        fleetsByProvince.set(army.province, byNation)
      }

      byNation.set(army.nation, addF32(byNation.get(army.nation) ?? 0, army.hitPoints))
    }

    for (let province = 0; province < this.world.provinces.count; province++) {
      if (!this.isCoastal(province)) continue

      const owner = this.world.provinces.controller[province]!
      if (owner === NO_OWNER) continue

      this.#evaluateCrossings(province, owner, fleetsByProvince)
    }
  }

  applyDailyEffects(): void {
    const morale = this.world.provinces.morale

    for (let province = 0; province < this.world.provinces.count; province++) {
      if (this.#blockaded[province] === 0) continue

      morale[province] = Math.max(MORALE_FLOOR, subF32(morale[province]!, MORALE_PENALTY_PER_DAY))
    }
  }

  blockadedCountOf(nation: number): number {
    let count = 0

    for (let province = 0; province < this.world.provinces.count; province++) {
      if (this.#blockaded[province] !== 0 && this.world.provinces.controller[province] === nation) {
        count++
      }
    }

    return count
  }

  /*
   * Map, bukan objek biasa: urutan sisipnya adalah urutan yang sama yang
   * dijalani Dictionary rujukan, dan penjumlahan float tidak asosiatif sehingga
   * mengubah urutannya mengubah hasilnya. Seri kekuatan juga jatuh ke armada
   * yang tercatat lebih dulu, bukan ke indeks negara terkecil.
   */
  #evaluateCrossings(
    province: number,
    owner: number,
    fleetsByProvince: ReadonlyMap<number, ReadonlyMap<number, number>>,
  ): void {
    const hostile = new Map<number, number>()
    let friendly = 0

    for (const crossing of this.sea.neighboursOf(province)) {
      const byNation = fleetsByProvince.get(crossing)
      if (byNation === undefined) continue

      for (const [nation, strength] of byNation) {
        if (nation === owner) {
          friendly = addF32(friendly, strength)
        } else if (this.relations.atWar(nation, owner)) {
          hostile.set(nation, addF32(hostile.get(nation) ?? 0, strength))
        }
      }
    }

    if (hostile.size === 0) return

    let strongest = 0
    let best = 0
    for (const [nation, strength] of hostile) {
      if (strength > best) {
        best = strength
        strongest = nation
      }
    }

    if (best < mulF32(Math.max(friendly, 1), STRENGTH_RATIO)) return

    this.#blockaded[province] = 1
    this.#blockader[province] = strongest
    this.#reports.push({
      province,
      blockader: strongest,
      victim: owner,
      fleetStrength: Math.trunc(best),
    })
  }
}

function carriesFleet(army: Army): boolean {
  for (const unit of army.units) {
    if (unit.unitClass.domain === Domain.Sea) return true
  }

  return false
}

File: sim/NationRise.Core/Military/Blockade.cs
using NationRise.Core.Data;
using NationRise.Core.Diplomacy;
using NationRise.Core.World;

namespace NationRise.Core.Military;

public readonly record struct BlockadeReport(int Province, ushort Blockader, ushort Victim, int FleetStrength);

/*
   Fleets sitting on a coastal province's sea links cut it off from the world.
   On an archipelago this is the only encirclement that exists: the sea graph
   is dense enough that no single land province is a chokepoint, so a siege of
   Indonesia has to be fought at sea or not at all.
*/
public sealed class Blockade(WorldState world, Relations relations, ProvinceGraph sea)
{
    /* A blockading fleet must outweigh whatever defends the crossing, or a
       single patrol boat could strangle a coastline. */
    public const float StrengthRatio = 2.0f;

    public const float ProductionPenalty = 0.50f;
    public const float MoralePenaltyPerDay = 0.03f;
    public const int SiegeProgressPerDay = 3;

    private readonly bool[] _blockaded = new bool[world.Provinces.Count];
    private readonly ushort[] _blockader = new ushort[world.Provinces.Count];
    private readonly List<BlockadeReport> _reports = [];

    public IReadOnlyList<BlockadeReport> Reports => _reports;

    public bool IsBlockaded(int province) => _blockaded[province];

    public ushort BlockaderOf(int province) => _blockader[province];

    public float ProductionMultiplierFor(int province) =>
        _blockaded[province] ? 1f - ProductionPenalty : 1f;

    /* Coastal means it has a sea crossing at all. An inland province cannot be
       blockaded no matter how strong the fleet, which is why holding the
       interior is worth something on an archipelago. */
    public bool IsCoastal(int province) => sea.NeighboursOf(province).Length > 0;

    public void Recompute(IReadOnlyDictionary<int, Army> armies)
    {
        Array.Clear(_blockaded);
        Array.Fill(_blockader, ProvinceStore.NoOwner);
        _reports.Clear();

        var fleetsByProvince = new Dictionary<int, Dictionary<ushort, float>>();

        foreach (Army army in armies.Values)
        {
            if (army.IsDestroyed || !CarriesFleet(army))
            {
                continue;
            }

            if (!fleetsByProvince.TryGetValue(army.Province, out Dictionary<ushort, float>? byNation))
            {
                byNation = [];
                fleetsByProvince[army.Province] = byNation;
            }

            byNation[army.Nation] = byNation.GetValueOrDefault(army.Nation) + army.HitPoints;
        }

        for (int province = 0; province < world.Provinces.Count; province++)
        {
            if (!IsCoastal(province))
            {
                continue;
            }

            ushort owner = world.Provinces.Controller[province];
            if (owner == ProvinceStore.NoOwner)
            {
                continue;
            }

            EvaluateCrossings(province, owner, fleetsByProvince);
        }
    }

    private void EvaluateCrossings(
        int province,
        ushort owner,
        Dictionary<int, Dictionary<ushort, float>> fleetsByProvince)
    {
        var hostile = new Dictionary<ushort, float>();
        float friendly = 0f;

        foreach (ushort crossing in sea.NeighboursOf(province))
        {
            if (!fleetsByProvince.TryGetValue(crossing, out Dictionary<ushort, float>? byNation))
            {
                continue;
            }

            foreach ((ushort nation, float strength) in byNation)
            {
                if (nation == owner)
                {
                    friendly += strength;
                }
                else if (relations.AtWar(nation, owner))
                {
                    hostile[nation] = hostile.GetValueOrDefault(nation) + strength;
                }
            }
        }

        if (hostile.Count == 0)
        {
            return;
        }

        ushort strongest = 0;
        float best = 0f;
        foreach ((ushort nation, float strength) in hostile)
        {
            if (strength > best)
            {
                best = strength;
                strongest = nation;
            }
        }

        if (best < MathF.Max(friendly, 1f) * StrengthRatio)
        {
            return;
        }

        _blockaded[province] = true;
        _blockader[province] = strongest;
        _reports.Add(new BlockadeReport(province, strongest, owner, (int)best));
    }

    private static bool CarriesFleet(Army army)
    {
        foreach (UnitInstance unit in army.Units)
        {
            if (unit.Class.Domain == Domain.Sea)
            {
                return true;
            }
        }

        return false;
    }

    public void ApplyDailyEffects()
    {
        for (int province = 0; province < world.Provinces.Count; province++)
        {
            if (_blockaded[province])
            {
                world.Provinces.Morale[province] =
                    MathF.Max(0.05f, world.Provinces.Morale[province] - MoralePenaltyPerDay);
            }
        }
    }

    public int BlockadedCountOf(ushort nation)
    {
        int count = 0;
        for (int province = 0; province < world.Provinces.Count; province++)
        {
            if (_blockaded[province] && world.Provinces.Controller[province] == nation)
            {
                count++;
            }
        }

        return count;
    }
}
